src/main/scala/org/broadinstitute/dsde/workbench/sam/azure/AzureService.scala

```diff
@@ -126,6 +136,79 @@ class AzureService(crlService: CrlService, directoryDAO: DirectoryDAO, azureMana
       createdPet <- directoryDAO.createPetManagedIdentity(petToCreate, samRequestContext)
     } yield (createdPet, true)
 
+  def getOrCreateActionManagedIdentity(
+      resource: FullyQualifiedResourceId,
+      resourceAction: ResourceAction,
+      billingProfileId: BillingProfileId,
+      samRequestContext: SamRequestContext
+  ): IO[(ActionManagedIdentity, Boolean)] = {
+    val id = ActionManagedIdentityId(resource, resourceAction, billingProfileId)
+    for {
+      existingAmiOpt <- directoryDAO.loadActionManagedIdentity(id, samRequestContext)
+      ami <- existingAmiOpt match {
+        // pet exists in Sam DB - return it
+        case Some(p) => IO.pure((p, false))
+        // pet does not exist in Sam DB - create it
+        case None => createActionManagedIdentity(id, samRequestContext)
+      }
+    } yield ami
+  }
+
+  def getActionManagedIdentity(
+      resource: FullyQualifiedResourceId,
+      resourceAction: ResourceAction,
+      samRequestContext: SamRequestContext
+  ): IO[Option[ActionManagedIdentity]] = directoryDAO.loadActionManagedIdentity(resource, resourceAction, samRequestContext)
+
+  private def createActionManagedIdentity(
+      id: ActionManagedIdentityId,
+      samRequestContext: SamRequestContext
+  ): IO[(ActionManagedIdentity, Boolean)] =
+    for {
+      mrgOpt <- azureManagedResourceGroupDAO.getManagedResourceGroupByBillingProfileId(id.billingProfileId, samRequestContext)
+      mrg <- IO.fromOption(mrgOpt)(
+        new WorkbenchExceptionWithErrorReport(
+          ErrorReport(StatusCodes.NotFound, s"Managed Resource Group with Billing Profile ID [${id.billingProfileId}] does not exist")
+        )
+      )
+      mrgCoordinates = mrg.managedResourceGroupCoordinates
+      // mapping the result of the validate call to ensure that validation happens before anything is created in Azure
+      validatedMrgCoordinates <- validateManagedResourceGroup(mrgCoordinates, samRequestContext).map(_ => mrg.managedResourceGroupCoordinates)
+      msiManager <- crlService.buildMsiManager(validatedMrgCoordinates.tenantId, validatedMrgCoordinates.subscriptionId)
```

Comment:
Just mentioning that `crlService.buildMsiManager` should be correctly Janitorized in BEEs/test envs. So action managed identities should get cleaned up in integration tests.

Reply:
Sweet!